File: my-folder/0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
		int zeroCount = 0;
		for (int i = 0; i < nums.length; i++) {
			if (nums[i] == 0) {
				if (++zeroCount == 2) {
					break;
				}
			}
		}
		if (zeroCount == 2) {
			nums = new int[nums.length];
		} else if (zeroCount == 1) {
			int product = 1;
			for (int i = 0; i < nums.length; i++) {
				if (nums[i] == 0) {
					continue;
				}
				product *= nums[i];
			}
			for (int i = 0; i < nums.length; i++) {
				if (nums[i] == 0) {
					nums[i] = product;
					continue;
				}
				nums[i] = 0;
			}
		} else {
			int product = 1;
			for (int i = 0; i < nums.length; i++) {
				product *= nums[i];
			}
			for (int i = 0; i < nums.length; i++) {
				nums[i] = product / nums[i];
			}
		}
		return nums;
	}
}

File: my-folder/0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res;
        res.push_back(1);
        for (int i = 1; i < n; i++)
        {
            res.push_back(res[i - 1] * nums[i - 1]);
        }
        int cur = 1;
        for (int i = n - 2; i > -1; i--)
        {
            cur *= nums[i + 1]; 
            res[i] *= cur;
        }
        return res;
    }
};
